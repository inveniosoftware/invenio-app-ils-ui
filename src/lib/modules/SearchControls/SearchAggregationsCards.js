import { BucketAggregation } from 'react-searchkit';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getSearchConfig } from '@config';

/**
 * Component wrapping BucketAggregation to provide custom display for
 * filter values
 *
 * if searchConfig provides property 'labels' for the filter
 * it will display the label mapped for current bucket key
 */
export default class SearchAggregationsCards extends Component {
  render() {
    const { modelName } = this.props;
    const searchConfig = getSearchConfig(modelName);
    if (searchConfig.FILTERS.length <= 0) {
      return <p>No filters available for this search.</p>;
    }

    return searchConfig.FILTERS.map((filter) => {
      return (
        <BucketAggregation
          key={filter.field}
          title={filter.title}
          agg={{
            field: filter.field,
            aggName: filter.aggName,
          }}
          overridableId="card"
        />
      );
    });
  }
}

SearchAggregationsCards.propTypes = {
  modelName: PropTypes.string.isRequired,
};
