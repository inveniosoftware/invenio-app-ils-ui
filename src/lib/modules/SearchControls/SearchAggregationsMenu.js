import _isEmpty from 'lodash/isEmpty';
import { Accordion, Dropdown } from 'semantic-ui-react';
import { BucketAggregation } from 'react-searchkit';
import React, { Component } from 'react';
import { getSearchConfig } from '@config';
import PropTypes from 'prop-types';

export default class SearchAggregationsMenu extends Component {
  constructor(props) {
    super(props);
    this.state = { menuOpen: false };
  }

  toggleMenuOpen = (e, elem) => {
    const { menuOpen } = this.state;
    this.setState({ menuOpen: !menuOpen });
  };

  closeMenu = (e, elem) => {
    this.setState({ menuOpen: false });
  };

  render() {
    const { modelName } = this.props;
    const { menuOpen } = this.state;
    const searchConfig = getSearchConfig(modelName);
    // this element is needed to prevent the Dropdown menu
    // children from closing on click
    const trigger = (
      <div
        className="text"
        role="alert"
        aria-live="polite"
        aria-atomic="true"
        onClick={this.toggleMenuOpen}
      >
        Filters
      </div>
    );

    return (
      <Dropdown
        item
        upward={false}
        size="small"
        pointing="top left"
        className="link item mobile-filters-dropdown"
        onChange={() => {}}
        disabled={_isEmpty(searchConfig.FILTERS)}
        open={menuOpen}
        trigger={trigger}
        // closeOnBlur
        closeOnEscape
        onBlur={this.closeMenu}
        scrolling
      >
        <Dropdown.Menu key="sub-menu">
          <Dropdown.Header>Refine search</Dropdown.Header>
          <Accordion styled fluid>
            {searchConfig.FILTERS.map(filter => (
              <BucketAggregation
                key={filter.field}
                title={filter.title}
                agg={{ field: filter.field, aggName: filter.aggName }}
                overridableId="menu"
              />
            ))}
          </Accordion>
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

SearchAggregationsMenu.propTypes = {
  modelName: PropTypes.string.isRequired,
};
