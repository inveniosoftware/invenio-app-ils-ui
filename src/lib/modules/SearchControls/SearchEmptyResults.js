import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Header, Icon, Segment } from 'semantic-ui-react';
import ClearButton from '@modules/SearchControls/ClearButton';

export default class SearchEmptyResults extends Component {
  render() {
    const { queryString, resetQuery, extraContent } = this.props;
    const currentSearch = `Current search "${queryString}"`;
    return (
      <Segment placeholder textAlign="center">
        <Header icon>
          <Icon name="search" />
          No results found!
        </Header>
        {queryString && (
          <>
            <div className="empty-results-current">{currentSearch}</div>
            <Segment.Inline>
              <ClearButton clickHandler={resetQuery} /> {extraContent}
            </Segment.Inline>
          </>
        )}
      </Segment>
    );
  }
}

SearchEmptyResults.propTypes = {
  queryString: PropTypes.string,
  resetQuery: PropTypes.func.isRequired,
  extraContent: PropTypes.node,
};

SearchEmptyResults.defaultProps = {
  queryString: '',
  extraContent: null,
};
