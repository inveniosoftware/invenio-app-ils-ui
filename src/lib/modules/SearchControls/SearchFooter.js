import { Media } from '@components/Media';
import { Grid } from 'semantic-ui-react';
import React, { Component } from 'react';
import SearchPagination from './SearchPagination';

export default class SearchFooter extends Component {
  render() {
    return (
      <Media greaterThan="mobile">
        <Grid textAlign="center" className="search-footer-pagination">
          <Grid.Column>
            <SearchPagination />
          </Grid.Column>
        </Grid>
      </Media>
    );
  }
}
