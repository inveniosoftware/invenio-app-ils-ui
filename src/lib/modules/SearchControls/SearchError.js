import { getStaticPageByName } from '@config';
import { SearchMessage } from '@modules/SearchControls/SearchMessage';
import _get from 'lodash/get';
import PropTypes from 'prop-types';
import React from 'react';
import Overridable from 'react-overridable';
import { Link } from 'react-router-dom';
import { Header } from 'semantic-ui-react';

export const SearchError = ({ error }) => {
  if (_get(error, 'response.status') === 400) {
    return (
      <SearchMessage title="Unrecognized search query." icon="question">
        The query you entered was not recognized.
      </SearchMessage> // When implemented, add link to documentation
    );
  } else {
    return (
      <Overridable id="SearchError.layout">
        <SearchMessage
          title="Something went wrong while fetching results."
          icon="warning"
        >
          <Header.Subheader>
            Please try again later. If the problem persists,{' '}
            <Link to={getStaticPageByName('contact').route}>contact us</Link>.
          </Header.Subheader>
        </SearchMessage>
      </Overridable>
    );
  }
};

SearchError.propTypes = {
  error: PropTypes.object.isRequired,
};
