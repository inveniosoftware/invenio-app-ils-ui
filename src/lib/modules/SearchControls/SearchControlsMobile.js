import { BucketAggregation, Count, SortBy } from 'react-searchkit';
import React, { Component } from 'react';
import { Grid, Container, Menu, Sticky } from 'semantic-ui-react';
import SearchResultsPerPage from './SearchResultsPerPage';
import SearchAggregationsMenu from './SearchAggregationsMenu';
import SearchSortOrder from './SearchSortOrder';
import PropTypes from 'prop-types';
import { getSearchConfig } from '@config';
import find from 'lodash/find';

export class SearchControlsMobile extends Component {
  renderCount = totalResults => {
    return (
      <div className="search-results-counter">{totalResults} results found</div>
    );
  };

  render() {
    const { stickyRef, modelName, withSortOrder } = this.props;
    const searchConfig = getSearchConfig(modelName);
    const loanFilter = find(
      searchConfig.FILTERS,
      o => o.aggName === 'availability'
    );
    return (
      <Container fluid className="mobile-search-controls">
        <Sticky context={stickyRef} offset={66}>
          <Container fluid className="fs-search-controls-mobile">
            <Menu fluid borderless>
              <Menu.Item>
                <BucketAggregation
                  key={loanFilter.field}
                  title={loanFilter.title}
                  agg={{ field: loanFilter.field, aggName: loanFilter.aggName }}
                  overridableId="available-for-loan"
                />
              </Menu.Item>
              <SearchAggregationsMenu modelName={modelName} />
              {searchConfig.SORT_BY.length > 0 ? (
                <SortBy
                  values={searchConfig.SORT_BY}
                  defaultValue={searchConfig.SORT_BY[0].value}
                  defaultValueOnEmptyString={
                    searchConfig.SORT_BY_ON_EMPTY_QUERY
                  }
                  overridableId="mobile"
                />
              ) : null}
              {withSortOrder && <SearchSortOrder modelName={modelName} />}
            </Menu>
            <Container>
              <Grid columns={2}>
                <Grid.Column width={8} className="vertical-align-content">
                  <div>
                    <Count
                      label={cmp => (
                        <div className="mobile-count">{cmp} results found</div>
                      )}
                    />
                  </div>
                </Grid.Column>
                <Grid.Column
                  width={8}
                  className="vertical-align-content"
                  textAlign="right"
                >
                  <div>
                    <SearchResultsPerPage
                      modelName={modelName}
                      label={cmp => (
                        <div className="mobile-results-page">
                          {cmp} results per page
                        </div>
                      )}
                    />
                  </div>
                </Grid.Column>
              </Grid>
            </Container>
          </Container>
        </Sticky>
      </Container>
    );
  }
}

SearchControlsMobile.propTypes = {
  modelName: PropTypes.string.isRequired,
  stickyRef: PropTypes.object,
  withSortOrder: PropTypes.bool,
};

SearchControlsMobile.defaultProps = {
  stickyRef: null,
  withSortOrder: true,
};
