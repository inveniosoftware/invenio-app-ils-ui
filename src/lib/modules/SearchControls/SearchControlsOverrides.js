import { Pagination } from '@components/Pagination';
import SearchBar from '@components/SearchBar/SearchBar';
import { MenuBucketAggregationValueElementOverrides } from '@modules/SearchControls/overridden/MenuBucketAggregationElement';
import { AvailableLoanBucketAggregationValues } from './overridden/AvailableLoanBucketAggregationValues';
import { AvailableLoanBucketAggregationElement } from './overridden/AvailableLoanBucketAggregationElement';
import { SortByElement } from './overridden/SearchSortByElement';
import { SortByElementMobile } from './overridden/SearchSortByElementMobile';
import SearchResultsList from './SearchResultsList';
import SearchEmptyResults from './SearchEmptyResults';
import { CardBucketAggregationValueElementOverrides } from './overridden/CardBucketAggregationValueElement';
import { SearchError } from '@modules/SearchControls/SearchError';

export const SearchControlsOverridesMap = {
  'BucketAggregationValues.element.card': CardBucketAggregationValueElementOverrides,
  'BucketAggregation.element.menu': MenuBucketAggregationValueElementOverrides,
  'BucketAggregation.element.available-for-loan': AvailableLoanBucketAggregationElement,
  'BucketAggregationValues.element.available-for-loan': AvailableLoanBucketAggregationValues,
  SearchBar: SearchBar,
  'Pagination.element': Pagination,
  'EmptyResults.element': SearchEmptyResults,
  'Error.element': SearchError,
  ResultsList: SearchResultsList,
  'SortBy.element.mobile': SortByElementMobile,
  'SortBy.element.desktop': SortByElement,
};
