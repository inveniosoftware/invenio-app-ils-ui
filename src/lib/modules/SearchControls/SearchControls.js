import { Count, LayoutSwitcher } from 'react-searchkit';
import React, { Component } from 'react';
import { Grid, Responsive } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import SearchSortBy from './SearchSortBy';
import SearchSortOrder from './SearchSortOrder';
import SearchResultsPerPage from './SearchResultsPerPage';
import { SearchControlsMobile } from './SearchControlsMobile';
import SearchPagination from './SearchPagination';

export class SearchControls extends Component {
  render() {
    const {
      withLayoutSwitcher,
      defaultLayout,
      modelName,
      withSortOrder,
    } = this.props;
    return (
      <>
        <Responsive {...Responsive.onlyComputer}>
          <Grid columns={3} className="search-controls">
            <Grid.Column largeScreen={5} computer={6}>
              <Grid columns={2}>
                {withLayoutSwitcher && (
                  <Grid.Column width={6} className="layout-switcher">
                    <LayoutSwitcher defaultLayout={defaultLayout} />
                  </Grid.Column>
                )}
                <Grid.Column width={10}>
                  <Count label={cmp => <div>{cmp} results found</div>} />
                  <SearchResultsPerPage modelName={modelName} />
                </Grid.Column>
              </Grid>
            </Grid.Column>
            <Grid.Column
              largeScreen={6}
              computer={4}
              textAlign="center"
              className="search-pagination-column"
            >
              <Responsive minWidth={Responsive.onlyLargeScreen.minWidth}>
                <SearchPagination />
              </Responsive>
            </Grid.Column>
            <Grid.Column
              textAlign="right"
              largeScreen={5}
              computer={6}
              className="search-sort-options-column"
            >
              <div className="sort-by-filters">
                <SearchSortBy modelName={modelName} />
                {withSortOrder && <SearchSortOrder modelName={modelName} />}
              </div>
            </Grid.Column>
          </Grid>
        </Responsive>
        <Responsive maxWidth={Responsive.onlyLargeScreen.minWidth - 1}>
          <Grid>
            <Grid.Column width={16} textAlign="center">
              <SearchPagination />
            </Grid.Column>
          </Grid>
        </Responsive>
        <Responsive maxWidth={Responsive.onlyTablet.maxWidth}>
          <SearchControlsMobile
            modelName={modelName}
            withSortOrder={withSortOrder}
          />
        </Responsive>
      </>
    );
  }
}

SearchControls.propTypes = {
  modelName: PropTypes.string.isRequired,
  withLayoutSwitcher: PropTypes.bool,
  defaultLayout: PropTypes.oneOf(['grid', 'list']),
  withSortOrder: PropTypes.bool,
};

SearchControls.defaultProps = {
  withLayoutSwitcher: true,
  defaultLayout: 'grid',
  withSortOrder: true,
};
