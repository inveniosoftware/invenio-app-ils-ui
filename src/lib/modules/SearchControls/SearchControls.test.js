import React from 'react';
import { shallow } from 'enzyme';
import * as testData from '@testData/documents.json';

import { SearchControls } from './SearchControls';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const data = {
  hits: {
    total: 2,
    hits: [{ metadata: testData[0] }, { metadata: testData[1] }],
  },
};

let store;
beforeEach(() => {
  store = mockStore({
    isLoading: false,
    data: data,
    hasError: false,
  });
  store.clearActions();
});

describe('SearchControls tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  it('should load the SearchControls component', () => {
    const component = shallow(<SearchControls modelName="DOCUMENTS" />);
    expect(component).toMatchSnapshot();
  });
});
