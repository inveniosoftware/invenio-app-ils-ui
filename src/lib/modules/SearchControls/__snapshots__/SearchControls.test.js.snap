// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchControls tests should load the SearchControls component 1`] = `
<Fragment>
  <Responsive
    getWidth={[Function]}
    minWidth={992}
  >
    <Grid
      className="search-controls"
      columns={3}
    >
      <GridColumn
        computer={5}
        largeScreen={5}
      >
        <Grid
          columns={2}
        >
          <GridColumn
            className="layout-switcher"
            width={6}
          >
            <Connect(Overridable(LayoutSwitcher))
              defaultLayout="grid"
            />
          </GridColumn>
          <GridColumn
            width={10}
          >
            <Connect(Overridable(Count))
              label={[Function]}
            />
            <SearchResultsPerPage
              modelName="DOCUMENTS"
            />
          </GridColumn>
        </Grid>
      </GridColumn>
      <GridColumn
        className="search-pagination-column"
        computer={6}
        largeScreen={6}
        textAlign="center"
      >
        <Responsive
          getWidth={[Function]}
          minWidth={1200}
        >
          <SearchPagination />
        </Responsive>
      </GridColumn>
      <GridColumn
        className="search-sort-options-column"
        computer={5}
        largeScreen={5}
        textAlign="right"
      >
        <div
          className="sort-by-filters"
        >
          <SearchSortBy
            modelName="DOCUMENTS"
          />
        </div>
      </GridColumn>
    </Grid>
  </Responsive>
  <Responsive
    getWidth={[Function]}
    maxWidth={1199}
  >
    <Grid>
      <GridColumn
        textAlign="center"
        width={16}
      >
        <SearchPagination />
      </GridColumn>
    </Grid>
  </Responsive>
  <Responsive
    getWidth={[Function]}
    maxWidth={991}
  >
    <SearchControlsMobile
      modelName="DOCUMENTS"
      stickyRef={null}
    />
  </Responsive>
</Fragment>
`;
